go/src/koding/klient/machine/mount/index/index.go

```diff
@@ -0,0 +1,328 @@
+package index
+
+import (
+	"crypto/sha1"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+// Entry represents a single file registered to index.
+type Entry struct {
+	Name  string      `json:"name"`  // The relative name of the file.
+	CTime int64       `json:"ctime"` // Metadata change time since EPOCH.
+	MTime int64       `json:"mtime"` // File data change time since EPOCH.
+	Mode  os.FileMode `json:"mode"`  // File mode and permission bits.
+	Size  int64       `json:"size"`  // Size of the file.
+	SHA1  []byte      `json:"sha1"`  // SHA-1 hash of file content.
+}
+
+// NewEntryFile creates new Entry from a file stored under path argument.
+// Info is optional and, if given, should store LStat result on the given file.
+func NewEntryFile(root, path string, info os.FileInfo) (e *Entry, err error) {
+	if info == nil {
+		if info, err = os.Lstat(path); err != nil {
+			return nil, err
+		}
+	}
+
+	// Get relative file name.
+	name, err := filepath.Rel(root, path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute file's SHA-1 sum.
+	var sum []byte
+	if !info.IsDir() {
+		if sum, err = readSHA1(path); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Entry{
+		Name:  filepath.ToSlash(name),
+		CTime: ctimeFromSys(info),
+		MTime: info.ModTime().UnixNano(),
+		Mode:  info.Mode(),
+		Size:  info.Size(),
+		SHA1:  sum,
+	}, nil
+}
+
+// readSHA1 computes SHA-1 sum from a given file content.
+func readSHA1(path string) ([]byte, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	hash := sha1.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return nil, err
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// ChangeMeta indicates what change has been done on a given file.
+type ChangeMeta uint32
+
+const (
+	ChangeMetaUpdate ChangeMeta = 1 << iota // File was updated.
+	ChangeMetaRemove                        // File was removed.
+	ChangeMetaAdd                           // File was added.
+
+	ChangeMetaLarge ChangeMeta = 1 << (8 + iota) // File size is above 4GB.
+)
+
+// Change describes single file change.
+type Change struct {
+	Name      string     `json:"name"`      // The relative name of the file.
+	Size      uint32     `json:"size"`      // Size of the file truncated to 32 bits.
+	Meta      ChangeMeta `json:"meta"`      // The type of operation made on file entry.
+	CreatedAt int64      `json:"createdAt"` // Change creation time since EPOCH.
+}
+
+// ChangeSlice stores multiple changes.
+type ChangeSlice []Change
+
+func (cs ChangeSlice) Len() int           { return len(cs) }
+func (cs ChangeSlice) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
+func (cs ChangeSlice) Less(i, j int) bool { return cs[i].Name < cs[j].Name }
+
+// Index stores a virtual working tree state. It recursively records objects in
+// a given root path and allows to efficiently detect changes on it.
+type Index struct {
```

Comment:
A: Everything is possible:). But, do we want one index for all mounts? It is guarded by mutex so eg. unmounting will stop all other mounts making it slow.

Other reason is that when index get broken we can just drop it and recreate. It's better to make this for a single mount rather than all of them(some of other ones might be even unreachable at that moment - machine is powered off).